Testing!

sco1 opened the pull request:
This PR adds testing, yay! This is my first foray into formal testing with Python so please help me be sure I'm not doing something completely terrible.

Here are the currently targeted points for testing:

- [x] Function parsing
- [x] Argument parsing
- [x] Classification of Function as fully annotated
- [x] Warning classification
- [x] Emitting correct line & column numbers
- [x] Formatting warning message with argument name (where relevant)
- [x] Outputting correct tuple for flake8
- [x] `Function` and `Argument` object `__str__` and `__repr__` methods

Additionally:
* Adds testing to the CI & a blurb in the README
* Adds coverage reporting as a pipenv script.
* Adds coverage reporting to CI
* Replaces the debug string methods for `Annotation` and `Function` with proper `__str__` and `__repr__` methods
* Adds a [Dependabot](https://dependabot.com/) configuration file to limit the scope of the dependency checking to the non-dev dependences. I have enabled Dependabot for this project in order to help with keeping on top of changes to flake8.
* Fixes incorrectly configured flake8-docstrings (see: https://github.com/python-discord/organisation/issues/131)

See: #7 
Closes: #16
Closes: #17
Closes: #19

MarkKoz commented on testing/test_column_line_numbers.py:
```diff
@@ -0,0 +1,57 @@
+# (line, column) tuples where we should get linting errors in the test file
+# Line numbers are 1-indexed
+# Column offsets are 0-indexed when yielded by our checker; flake8 adds 1 when emitted
+SHOULD_ERROR = ((16, 8), (16, 11), (24, 4), (25, 4), (26, 2))
```
This is a fragile way of doing things but I can't think of other ways.

sco1 replied:
This mirrors what other plugins do (e.g. [bugbear](https://github.com/PyCQA/flake8-bugbear/tree/master/tests)) but I don't disagree that it's fragile. It's probably worth adding a note to the testing source that these values need to be adjusted if the source changes. This might be another good spot for a namedtuple too.

MarkKoz commented on testing/test_column_line_numbers.py:
```diff
@@ -0,0 +1,57 @@
+    tree = ast.parse(src)
+    lines = src.splitlines()
+
+    return list(checker.TypeHintChecker(tree, lines).run())
```
Why isn't this left as a generator? Does pytest use the same instance of a fixture for all tests?

sco1 replied:
I don't know. I think it's an artifact from an earlier (fixtureless) approach that I never changed.

MarkKoz commented on testing/test_column_line_numbers.py:
```diff
@@ -0,0 +1,57 @@
+    Note: Line numbers are 1-indexed
+    """
+    should_error_lines = [coordinate_pair[0] for coordinate_pair in SHOULD_ERROR]
+    raised_error_lines = [error_tuple[0] for error_tuple in parsed_errors]
```
It strikes me that this could be accomplished more efficiently by iterating both in parallel. `any()` could be used or each comparison could be asserted.

sco1 replied:
Asserting each line/column would probably be easier to follow in the event that one fails than a potentially dense list of numbers

MarkKoz commented on testing/test_flake8_format.py:
```diff
@@ -0,0 +1,80 @@
+    check.is_instance(emitted_error[2], str)
+
+    # Fourth value should be a type (not an instance) and the same as TypeHintChecker
+    check.is_instance(emitted_error[3], type)
```
Isn't this redundant given the following check?

sco1 replied:
Technically yes, but I found that figuring out the reason for a yielded assert error here confusing with the single equal assert because it has two failure modes: the value being an instance rather than a type, and the value being of the wrong type. I think being explicit here allows the developer to more quickly identify where the issue is.

That said, I don't know if the former (instance vs. type) is actually important to flake8 but I will take a look.

MarkKoz replied:
IMO the equality check suffices, but it doesn't hurt to leave this in so really I don't mind.

MarkKoz commented on testing/test_variable_formatting.py:
```diff
@@ -0,0 +1,66 @@
+class TestArgumentFormatting:
+    """Testing class for containerizing parsed error codes & running the fixtured tests."""
+
+    with TEST_FILE.open("r") as f:
```
```suggestion
    with TEST_FILE.open("r", encoding="utf-8") as f:
```

MarkKoz commented on testing/test_variable_formatting.py:
```diff
@@ -0,0 +1,66 @@
+
+    batched_error_codes = {error_type: [] for error_type in TEST_ARG_NAMES.keys()}
+    for code, arg_name in error_codes:
+        batched_error_codes[code].append(arg_name)
```
For some reason I found this really hard to wrap my head around, as in what the code does as a whole. Yet individually it doesn't seem unreadable by any means. Anyway, I think this is clearer:

```py
from collections import defaultdict

batched_error_codes = defaultdict(list)

for error in checker.TypeHintChecker(tree, lines).run():
    code, arg_name = _simplify_error(error)
    batched_error_codes[code].append(arg_name)
```

If you're not a fan of importing `defaultdict` then `setdefault()` is an alternative.

sco1 commented on testing/test_classifier.py:
```diff
@@ -0,0 +1,70 @@
+        """
+        error_object = object_attributes.argument_classifications[request.param]
+        function_object = Function("ArgumentTest", 0, 0, None, request.param[0], request.param[2])
+        argument_object = Argument("TestArgument", 0, 0, request.param[3])
```
I agree. I looked at namedtuples originally but moved away from them for a reason I can't remember.

sco1 commented on testing/test_classifier.py:
```diff
@@ -0,0 +1,70 @@
+        """
+        error_object = object_attributes.argument_classifications[request.param]
+        function_object = Function("ArgumentTest", 0, 0, None, request.param[0], request.param[2])
+        argument_object = Argument("TestArgument", 0, 0, request.param[3])
```
I was doing the same thing writing it, which should have been a flag...

sco1 commented:
Testing for Function & Argument parsing has been added, so this should be ready for review

MarkKoz commented on testing/test_parser.py:
```diff
@@ -0,0 +1,156 @@
+        return truth_arguments, parsed_arguments, request.param
+
+    def test_argument_parsing(
+        self, argument_lists: Tuple[List[Argument], List[Argument], str]
```
Shouldn't these two be on separate lines?

MarkKoz commented on testing/test_parser.py:
```diff
@@ -0,0 +1,156 @@
+
+        Argument objects are provided as a tuple of (truth, source) lists
+        """
+        for truth_arg, parsed_arg in zip_longest(argument_lists[0], argument_lists[1]):
```
Just an idea. I don't know if it's necessarily clearer but it is shorter.
```suggestion
        for truth_arg, parsed_arg in zip_longest(*argument_lists[:2]):
```

MarkKoz commented on testing/test_parser.py:
```diff
@@ -0,0 +1,156 @@
+    tree = ast.parse(src)
+    lines = src.splitlines()
+    visitor = FunctionVisitor(lines)
+    visitor.visit(tree)
```
Perhaps a helper function is in order for this? If this pattern is done in multiple test modules then it can be defined in `__init__.py`.

MarkKoz commented on testing/test_parser.py:
```diff
@@ -0,0 +1,156 @@
+    visitor.visit(tree)
+
+    @pytest.fixture(params=parser_object_attributes.parsed_functions.keys())
+    def functions(self, request) -> Tuple[Function, Function]:  # noqa
```
This fixture is quite similar to the argument test's fixture. Is there a nice way to make this more generic so that both can use it? The code is really simple and short (it's basically a `next()` with a filter) so maybe it's not worth it.

sco1 commented:
> I'm starting to realise these really are more like integration tests than unit tests.

From my exploration of the internet I think we have the opposite. Here's a [source](http://softwaretestingfundamentals.com/) for definitions (no idea if it's a good one, but the internet seems to generally agree with the definitions):

> UNIT TESTING is a level of software testing where individual units/ components of a software are tested. The purpose is to validate that each unit of the software performs as designed.

> INTEGRATION TESTING is a level of software testing where individual units are combined and tested as a group. The purpose of this level of testing is to expose faults in the interaction between integrated units.

I tried to structure the tests so individual components are being tested rather than attempting to write a set of function definitions that covers all of the scenarios. From my initial exploration, the latter seemingly leads to a pretty incomprehensible set of repetitive definitions in order to cover the possible permutations of parameters we cover. This is why I split testing of the parser & classifier into two distinct tests, allowing for them to be much more clearly parameterized. Everything else I believe should be fairly cleanly unitized.

I'm not sure how much additional integration testing is needed beyond what's currently written. The parser should be fully tested, and the output emitted for flake8 has its own test as well.

> Also, any plans to run test coverage?

I've added coverage to the dev dependencies & provided a pipenv script. Whether we want it added to the CI would be a good discussion to have.

Currently we're sitting at 98%:

```
----------- coverage: platform win32, python 3.7.3-final-0 -----------
Name                                Stmts   Miss  Cover   Missing
-----------------------------------------------------------------
flake8_annotations\__init__.py        105      6    94%   46, 89, 137, 195, 233-234
flake8_annotations\checker.py          57      0   100%
flake8_annotations\enums.py            15      0   100%
flake8_annotations\error_codes.py      85      0   100%
-----------------------------------------------------------------
TOTAL                                 262      6    98%
```

Currently missing are:

* `Argument._debug_summary` (`46`)
  * Don't really care about testing this, but it's easy enough to cover
* `Function.is_fully_annotated` (`89`)
  * Will add
* `Function.from_function_node` (`137`)
  * Missing case for line number fallback. I don't know if we can ever encounter this but will figure out how to mock it anyway
* `Function._debug_summary` (`195`)
  * Don't really care about testing this, but it's easy enough to cover
* `FunctionVisitor.visit_AsyncFunctionDef` (`233-234`)
  * Will add. Functionally no different than `FunctionVisitor.visit_FunctionDef` but easy enough to cover

MarkKoz commented on flake8_annotations/__init__.py:
```diff
@@ -244,3 +261,17 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
         )
+
+    @classmethod
+    def parse_file(cls, src_filepath: Path) -> "FunctionVisitor":  # Need to quote for 3.6 compat
```
Is this used anywhere in this package or only for testing? If the latter, I think this should be in `testing/__init__.py`.

sco1 replied:
I don't mind these where they are. A few plugins (e.g. [bugbear](https://github.com/PyCQA/flake8-bugbear/blob/master/bugbear.py#L56)) have this method as a fallback as well. If I could figure out what scenario causes the lines/ast to not be provided by flake8 I'd add it as well, but until I can figure that out I don't know how to test it.

MarkKoz commented on flake8_annotations/checker.py:
```diff
@@ -33,6 +34,17 @@ def run(self) -> Generator[error_codes.Error, None, None]:
                 yield classify_error(function, arg).to_flake8()
 
+    @staticmethod
+    def load_file(src_filepath: Path) -> Tuple[ast.Module, List[str]]:
```
Is this used anywhere in this package or only for testing? If the latter, I think this should be in `testing/__init__.py`.

MarkKoz commented:
> I've added coverage to the dev dependencies & provided a pipenv script. Whether we want it added to the CI would be a good discussion to have.

I think it should be added to CI. At the least just a report and not necessarily any sort of threshold for passing. The report can be published with an Azure task but it should still be configured to output to stdout. Apparently with `pytest-cov` it can be done with `--cov-report term --cov-report public:html`. For the publishing task, you can take a look at snekbox.

MarkKoz commented on azure-pipelines.yml:
```diff
@@ -39,4 +39,14 @@ jobs:
\ No newline at end of file
+      testRunTitle: 'flake8-annotations Lint & Test Results'
+
+  - script: python3 -m pytest --cov-report xml:cov.xml --cov=flake8_annotations testing/
```
Does an xml report also print to normally to stdout? Isn't `--cov-report term-missing` needed for that? The reason it should be printed to stdout is because Azure has failed before in showing published reports so the console output is a fallback.

MarkKoz replied:
This also applies to the flake8 output if it isn't already being shown in stdout. If you're generating XML for flake8 then the only way I've found to get both XML and stdout is to simply run flake8 twice unfortunately.

sco1 replied:
Added terminal output for code coverage and removed XML reporting for the linter, it's close to useless since you have to click on each error just to see where it is.

MarkKoz commented on azure-pipelines.yml:
```diff
@@ -1,6 +1,6 @@
-- job: lint
-  displayName: 'Linting with flake8'
+- job: lint_and_test
+  displayName: 'Lint with flake8 & test with pytest'
```
I don't think it's useful to put such details in the name. It gets cut off anyway when viewing it on the Azure build page.
```suggestion
  displayName: 'Lint & Test'
```

sco1 commented on azure-pipelines.yml:
```diff
@@ -29,11 +29,27 @@ jobs:
   - script: python3 -m pip install pipenv && pipenv install --dev --deploy --system && python3 -m pip install flake8-formatter-junit-xml
```
I'm all for making things nuclear bomb proof, but I'm not really convinced that the added burdeon of us evaluating new releases of pipenv to make sure they still install a project is a worthwhile use of our time. I'm comfortable hedging on the likelihood of PyPA breaking pipenv so fundamentally being quite low.

I don't mind pinning the formatter in the pipfile, it would help for local testing anyway.

MarkKoz replied:
Well, I've experienced my fair share of pipenv bugs. Though I think most of them came from the use of newer pip versions which may happen if Microsoft updates the Ubuntu agent we use. They still keep up with the latest Python versions clearly since the agent currently uses 3.7.4. I don't know if they update pip when they update Python.

> I don't mind pinning the formatter in the pipfile, it would help for local testing anyway.
It's fine to pin it without putting it in the Pipfile (by specifying the version in the install command here).

MarkKoz replied:
On that note we may be screwed anyway if a new Python version and new pip version are added to the agent. It doesn't look like MS keeps around old patch versions of Python, so we couldn't pin to that.

sco1 replied:
I think this would be better hashed out at the organization level, since it would potentially affect most of our projects.

scragly replied:
Yeah, it sounds like the impact is a bit larger than the scope of just this project. Definitely worth discussing more in future though as if it's a serious concern, it's best to be addressed across all projects that could be impacted. I'll resolve for now, but consider making an issue ticket in org.

lemonsaurus commented:
I don't have time to test this today, but the code looks phenomenal, Mark has done a thorough review and you basically have 100% test coverage so I don't think this desperately needs more testing.

I would just merge this.

MarkKoz commented:
The whole integration vs unit tests thing was fleshed out more on Discord. I came to the conclusion that they aren't really integration tests. However, some unit tests involve more units/components than they're actually concerned with testing. This was not addressed in this PR but I suppose it can be fixed later as the tests still work as they are.
